### ISSUE TITLE
Exposing cache data

### ISSUE THREAD

**Issue by ktsn** [HerringtonDarkholme/gulp-progeny#22]
Hi, thank you for the great tool! I really enjoy using it :)

What do you think about exposing cache data [like gulp-cached](https://github.com/contra/gulp-cached#clearing-the-cache)?
I'm tackling a large project and want the initial build to be faster. My idea is re-using the cache of the previous gulp process with (re-)storing the in-memory caches as a file. This approach works well for `gulp-cached`, I want to apply this for `gulp-progeny` too. :slightly_smiling_face:

**Comment by HerringtonDarkholme** [comment]
Thanks for trying out this plugin, my friend ;)
Currently I don't have idea about reusing cache file. Because gulp progeny relies on gulp-watch to keep dependencies up to date. If we restoring caches from file, we have to check whether it is stale or not, and to do that we need to re-read all the file again. So the initial building time will be slow :(

Do you have any idea to make it even faster?

**Comment by ktsn** [comment]
Ah, I did not describe my problem correctly, sorry!
We can achieve fast initial build by restoring the gulp-cached's cache now like:

```js
const cached = require('gulp-cached')

// save caches
fs.writeFile('./.cache', JSON.stringify(cached.caches))

// load caches
fs.readFile('./.cache', (error, data) => {
  cached.caches = JSON.parse(data)
})
```

But gulp-progeny becomes cannot detect compiled files' dependencies in this case. I guess this is because [these caches](https://github.com/HerringtonDarkholme/gulp-progeny/blob/master/src/index.coffee#L7-L8) are lost by restarting gulp process.

So, if it is make sense, can we add some api to get/set these caches from user land?

**Comment by HerringtonDarkholme** [comment]
I think exposing cache is easy. But reusing caches might be tricky. For example, if A depends on B , and this is captured in cache file. Then A changed,  it depends on C. gulp restarts and reads the cache file and think A still depends on B. That is a bug.

**Comment by ktsn** [comment]
I'm still not sure the code well but wouldn't the dependencies of updated files be refreshed when we restart gulp since the files will be processed initially? In my idea, if we restore entire state for the plugin, it will behave as same as when the state was stored.

Let me describe the example in visual (each row indicates the state on the same time, X -> Y indicates X depends on Y):

Event | Target file<br>deps | In-memory<br>cache deps | File cache<br>deps
--- | --- | --- | ---
1. start gulp | A -> B | (none) | (none) | (none)
2. A is processed | A -> B | A -> B | (none)
3. save cache | A -> B | A -> B | A -> B
4. A changed<br>(and processed) | A -> C | A -> C | A -> B
5. restart gulp | A -> C | (none) | A -> B
6. load cache | A -> C | A -> B | A -> B
7. A is processed | A -> C | A -> C | A -> B

If we don't save the cache on 4, it indeed cause inconsistency between actual deps and cache deps on 6. But I think it would be processed by gulp-progeny and updated its dependencies on the initial processing as the file won't hit in (gulp-cached's) cache since it was updated.

Would I be still missing something?

**Comment by HerringtonDarkholme** [comment]
> If we don't save the cache on 4, it indeed cause inconsistency between actual deps and cache deps on 6.

Yes, that's exactly what I'm worried about.

> updated its dependencies on the initial processing as the file won't hit in (gulp-cached's) cache since it was updated.

This will require gulp-cached compute all changed files since last change. If I read the code right, gulp cached only computes file's md5. And no restoring mechanism is supported.

Do you mean, that you can restoring gulp-cached's md5 cache. And then reading all the file and pass through gulp-cached to filter only changed files. Then, changed files go to gulp-progeny to compute all the changed files and their dependencies.

So visually. Suppose we have A,B, C, D four files, A depends on B in last build. A and C get updated, A depends on C this build. And D depends on A for both builds.

```
A, B, C, D // all files
    ↓
gulp-cached // with restored cache, md5sum of last build
    ↓
   A, C  // passed through because md5sum changed last time
    ↓
gulp-progeny // with restored cache, deps of last build
   ↓
  A, C, D //  D added to stream because its dependence A gets updated.
```

**Comment by HerringtonDarkholme** [comment]
That's totally make sense. And I believe I have exposed cache in previous release but that got removed in some refactor. I will add it back.

**Comment by ktsn** [comment]
> Do you mean, that you can restoring gulp-cached's md5 cache. And then reading all the file and pass through gulp-cached to filter only changed files. Then, changed files go to gulp-progeny to compute all the changed files and their dependencies.

Yes, I mean that. 

Thanks for considering to add it :)

**Comment by HerringtonDarkholme** [comment]
published as 0.3.2,
now you can access cache data by `progeny.caches`. There is also `progeny.processedFileNames` which indicates all files have been seen by progeny.

**Comment by ktsn** [comment]
Thanks! I'll try it later.
